examples/define-syntax-rule.kl

```diff
@@ -1,47 +1,110 @@
 #lang "prelude.kl"
 
 (import (shift "prelude.kl" 1))
+(import (shift "defun.kl" 1))
 (import (shift "do.kl" 1))
 (import (shift "list-syntax.kl" 1))
 (import (shift "quasiquote.kl" 1))
 (import (shift "syntax.kl" 1))
+(import (shift "temporaries.kl" 1))
 
+-- (define-macro my-macro
+--   (lambda (stx) ...))
+-- or
+-- (define-macro (my-macro arg1 arg2 arg3)
+--   ...)
 (define-macros
-  ((define-macro
-     (lambda (stx)
-       (syntax-case stx
-         ((list (_ pattern body))
-          (syntax-case pattern
-           ((cons macro-name args)
+  ([define-macro
+    (lambda (stx)
+      (syntax-case stx
+        [(list (_ pattern body))
+         (syntax-case pattern
+           [(ident macro-name)
             (pure `(define-macros
-                     ((,macro-name (lambda (stx)
-                                     (syntax-case stx
-                                      ((list ,pattern)
-                                       ,body)))))))))))))))
+                     ([,macro-name
+                       ,body])))]
+           [(cons macro-name args)
+            (pure `(define-macros
+                     ([,macro-name
+                       (lambda (stx)
+                         (syntax-case stx
+                           [(list ,pattern)
+                            ,body]))])))])]))]))
+
+(define-macro m1
+  (lambda (stx)
+    (pure `(,'quote ,stx))))
+
+(define-macro (m2 arg)
+  (pure `(,'quote ,arg)))
+
+(example (m1 foo bar baz))  -- '(m1 foo bar baz)
+(example (m2 foo))          -- 'foo
+
+
```

Comment:
The following implementation is way more complicated than it needs to be. I wrote it first, and then I wrote `syntax-pmatch`. I then tried to rewrite `define-syntax-rule` in terms of `syntax-pmatch`, but I encountered #80 . So I'll leave it as-is for now.